### PR TITLE
Conform to new API specs - Closes #139

### LINKED ISSUE #139: Conform to new API specs
The new API specifics as described here:

https://app.swaggerhub.com/apis/LiskHQ/Lisk/1.0.5

Should be implemented into `lisk-js` before the 1.0.0 release to be prepared.

In order to be reviewed better and more structure this issue will be divided into:

- [x] Erase parseTransaction #239
- [x] Add first parameter GET/POST/PUT for sendRequest #240
- [x] Rewrite popsicle request and erase sendRequestPromise, changeRequest... #241
- [x] Rewrite API functions with new API method #242
- [x] Apart from default parameter, add options object to the API functions #243
- [x] Functions that use GET should be named accordingly (getAccount, getVoters...) #244
- [x] Add missing dapp API functions #290

### PR DESCRIPTION
**Pull request by Tosch110**
Closes #139

### CODE REVIEW

**Comment by Tosch110** [comment]
This is a first proposal. It will likely have some changes that should reflect the actual inplementation into Lisk 1.0 and which is still work in progress.

Also function naming conventions and arguments ordering will be under discussion and likely to be changed before merged.

File: lib/api/liskApi.js [hunk 1]
```diff
@@ -188,12 +193,7 @@ LiskAPI.prototype.sendRequest = function (requestType, options, callback) {
 				that.sendRequest(requestType, options, callback);
 			}, 1000);
 		} else {
-			var rejectAnswer = { success: false, error: error, message: 'could not create http request to any of the given peers' };
-
-			if (callback && (typeof callback === 'function')) {
-				callback(rejectAnswer);
-			}
-			return rejectAnswer;
+			return { success: false, error: error, message: 'could not create http request to any of the given peers' };
```
Comment (willclarktech) [inline lib/api/liskApi.js]: Capitalise message.

File: lib/api/liskApi.js [hunk 2]
```diff
@@ -268,7 +254,7 @@ LiskAPI.prototype.listActiveDelegates = function (limit, callback) {
 LiskAPI.prototype.listStandbyDelegates = function (limit, callback) {
 	var standByOffset = 101;
 
-	this.sendRequest('delegates/', { limit: limit, orderBy: 'rate:asc', offset: standByOffset}, function (result) {
+	this.sendRequest('delegates/', { limit: limit, sort: 'rate:asc', offset: standByOffset}, function (result) {
```
Comment (willclarktech) [inline lib/api/liskApi.js]: I don't see `limit` or `offset` in the API docs.

File: lib/api/liskApi.js [hunk 3]
```diff
@@ -428,18 +488,26 @@ LiskAPI.prototype.listMultisignatureTransactions = function (callback) {
...
 
 	var request = {
 		requestMethod: 'POST',
 		requestUrl: privateApi.getFullUrl.call(this) + '/peer/' + 'transactions',
 		nethash: this.nethash,
-		requestParams: { transaction }
+		requestParams: { transaction: transaction }
 	};
 
 	privateApi.doPopsicleRequest.call(this, request).then(function (result) {
```
Comment (willclarktech) [inline lib/api/liskApi.js]: Why don't we have a generic request method that abstracts some of this functionality for `POST` requests like we do for `GET` (or why isn't there one which covers both)?

File: lib/api/liskApi.js [hunk 4]
```diff
@@ -428,18 +488,26 @@ LiskAPI.prototype.listMultisignatureTransactions = function (callback) {
...
+ * @param callback
+ *
+ * @return API object
+ */
+
 LiskAPI.prototype.broadcastSignedTransaction = function (transaction, callback) {
 
 	var request = {
 		requestMethod: 'POST',
 		requestUrl: privateApi.getFullUrl.call(this) + '/peer/' + 'transactions',
```
Comment (willclarktech) [inline lib/api/liskApi.js]: This doesn't conform to API docs (`/transactions`).

File: lib/api/liskApi.js [hunk 5]
```diff
@@ -407,14 +393,88 @@ LiskAPI.prototype.sendLSK = function (recipient, amount, secret, secondSecret, c
...
+LiskAPI.prototype.multiSignatureSign = function (secret, transaction, callback) {
+	var signedTx = LiskJS.multisignature.signTransaction(transaction, secret);
+
+	this.broadcastSignedTransaction(signedTx, function (result) {
+		return callback(result);
+	});
+};
+
+/**
+ * @method multiSignatureSign
```
Comment (willclarktech) [inline lib/api/liskApi.js]: Incorrect method name.

File: lib/api/liskApi.js [hunk 6]
```diff
@@ -407,14 +393,88 @@ LiskAPI.prototype.sendLSK = function (recipient, amount, secret, secondSecret, c
 };
 
 /**
+ * @method multiSignatureSign
+ * @param transaction object
```
Comment (willclarktech) [inline lib/api/liskApi.js]: Inconsistent params ordering.

File: lib/api/liskApi.js [hunk 7]
```diff
@@ -428,18 +488,26 @@ LiskAPI.prototype.listMultisignatureTransactions = function (callback) {
  */
 
 LiskAPI.prototype.getMultisignatureTransaction = function (transactionId, callback) {
-	this.sendRequest('transactions/multisignatures/get', { id: transactionId }, function (result) {
+	this.sendRequest('transactions/unsigned', { id: transactionId }, function (result) {
```
Comment (willclarktech) [inline lib/api/liskApi.js]: This does not conform to API docs.

File: lib/api/privateApi.js [hunk 8]
```diff
@@ -277,14 +211,19 @@ function doPopsicleRequest (requestValue) {
...
+	var requestValues = {};
+
+	requestValues.requestMethod = 'GET';
+	requestValues.requestUrl = getFullUrl.call(this) + '/api/' + requestType;
+
+	if (Object.keys(options).length > 0) {
+		requestValues.requestUrl = requestValues.requestUrl + serialiseHttpData.call(this, options, requestValues.requestMethod);
 	}
+
+	requestValues.requestParams = options;
```
Comment (willclarktech) [inline lib/api/privateApi.js]: Why are we doing query and params with the same keys and values?

File: lib/api/utils.js [hunk 9]
```diff
@@ -5,15 +5,15 @@ module.exports = {
  *
  * @return trimmed string
  */
-	trimObj: function rimObj(obj) {
+	trimObj: function trimIt(obj) {
```
Comment (willclarktech) [inline lib/api/utils.js]: This was already fixed to be named `trimObj` in 0ac3ad0

File: lib/api/utils.js [hunk 10]
```diff
@@ -27,9 +27,7 @@ module.exports = {
 		var parts = [];
 
 		for (var i in obj) {
-			if (obj.hasOwnProperty(i)) {
```
Comment (willclarktech) [inline lib/api/utils.js]: Prefer `for (var i of obj)` without the `obj.hasOwnProperty` check.

File: lib/transactions/multisignature.js [hunk 11]
```diff
@@ -88,7 +88,7 @@ function signTransaction (trs, secret) {
  * @return {Object}
  */
 
-function createMultisignature (secret, secondSecret, keysgroup, lifetime, min) {
+function createMultisignature (secret, keysgroup, lifetime, min, secondSecret) {
```
Comment (willclarktech) [inline lib/transactions/multisignature.js]: yuidoc comments need to be updated

File: lib/utils/mnemonic.js [hunk 12]
```diff
@@ -41,7 +41,7 @@ var Buffer = require('buffer/').Buffer;
  */
 
 function generateEmptyBytes(length) {
-	return Array.apply(null, Array(length)).map(x => 0); //eslint-disable-line
+	return Array.apply(null, Array(length)).map(function (x) { return 0 } ); //eslint-disable-line
```
Comment (willclarktech) [inline lib/utils/mnemonic.js]: `return new Array(length).fill(0)`

File: package.json [hunk 13]
```diff
@@ -27,6 +27,7 @@
     "chai": "=4.0.0",
     "chai-http": "=3.0.0",
     "coveralls": "=2.13.1",
+    "express": "^4.15.3",
```
Comment (willclarktech) [inline package.json]: What for?

File: test/mocha.opts [hunk 14]
```diff
@@ -1 +1,2 @@
---timeout 10000
\ No newline at end of file
+--timeout 10000
+--recursive test/**/*.js
```
Comment (willclarktech) [inline test/mocha.opts]: Why is this needed?

File: test/transactions/multisignature.js [hunk 15]
```diff
@@ -138,11 +138,9 @@ describe('multisignature.js', function () {
...
 
-			console.log(msigTransaction2);
-
 		});
 
 
 	});
 
-});
+});
```
Comment (willclarktech) [inline test/transactions/multisignature.js]: Prefer newline at ends of files.
